idx/metric_definition.go

```diff
@@ -0,0 +1,552 @@
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag key: %v", err))
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		panic(fmt.Sprintf("idx: Failed to retrieve interned tag value: %v", err))
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func (t TagKeyValue) createTagKeyValue() string {
+	bld := strings.Builder{}
+
+	key, err := IdxIntern.GetStringFromPtr(t.Key)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag key: ", err)
+		internError.Inc()
+	}
+	val, err := IdxIntern.GetStringFromPtr(t.Value)
+	if err != nil {
+		log.Error("idx: Failed to retrieve interned tag value: ", err)
+		internError.Inc()
+	}
+
+	bld.WriteString(key)
+	bld.WriteString("=")
+	bld.WriteString(val)
+
+	return bld.String()
+}
+
+func parseTagKeyValue(tag string) TagKeyValue {
+	var tkv TagKeyValue
+	if strings.Contains(tag, ";") {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	eqPos := strings.Index(tag, "=")
+	if eqPos < 0 {
+		log.Errorf("idx: Tag %q has an invalid format, ignoring", tag)
+		invalidTag.Inc()
+		return TagKeyValue{}
+	}
+	key, err := IdxIntern.AddOrGet([]byte(tag[:eqPos]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Key = key
+
+	value, err := IdxIntern.AddOrGet([]byte(tag[eqPos+1:]), true)
+	if err != nil {
+		log.Errorf("idx: Failed to intern tag %q, %v", tag, err)
+		internError.Inc()
+		return TagKeyValue{}
+	}
+	tkv.Value = value
+	return tkv
+}
+
+type KeyValuesSlice []TagKeyValue
+
+func (kvs KeyValuesSlice) Len() int      { return len(kvs) }
+func (kvs KeyValuesSlice) Swap(i, j int) { kvs[i], kvs[j] = kvs[j], kvs[i] }
+func (kvs KeyValuesSlice) Less(i, j int) bool {
+	k1, _ := IdxIntern.GetStringFromPtr(kvs[i].Key)
+	k2, _ := IdxIntern.GetStringFromPtr(kvs[j].Key)
+	return k1 < k2
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues struct {
+	KeyValues KeyValuesSlice
+}
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t *TagKeyValues) Strings() []string {
+	tags := make([]string, len((*t).KeyValues))
+	for i, tag := range (*t).KeyValues {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+func (t *TagKeyValues) ExtensionType() int8 {
+	return 90
+}
+
+func (t *TagKeyValues) Len() int {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		total += len(kv) + 1
+	}
+	return total
+}
+
+func (t *TagKeyValues) MarshalBinaryTo(b []byte) error {
+	var total int
+	for _, kv := range t.Strings() {
+		if strings.HasPrefix(kv, "name") {
+			continue
+		}
+		if kv == "=" || kv == "" {
```

Comment:
these values would indicate that an invalid tag has been parsed by `parseTagKeyValue()`. I think we should decide if the `TagKeyValues` struct needs to do input validation or not. If it does then the parsing function should do the validation, if not then the Marshaling function should also marshal invalid tags. But just checking for some of the possible error cases when marshaling seems a bit arbitrary.

Reply:
I think ideally we don't want any of these functions to be error checking the tags at all. These should all be checked at the point of ingest, and then never again. Even if we move our validation to one of the other applications that we typically use with Metrictank I think we still need to validate upon ingest inside Metrictank. That's why a shared validation library makes the most sense still.